runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/SingletonDockerContainerManager.java

```diff
@@ -0,0 +1,111 @@
+
+  private final GrpcFnServer<SdkHarnessClientControlService> controlServiceServer;
+  private final GrpcFnServer<GrpcLoggingService> loggingServiceServer;
+  private final GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer;
+  private final GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer;
+
+  private SingletonDockerContainerManager(
+      GrpcFnServer<SdkHarnessClientControlService> controlServiceServer,
+      GrpcFnServer<GrpcLoggingService> loggingServiceServer,
+      GrpcFnServer<ArtifactRetrievalService> retrievalServiceServer,
+      GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer) {
+    this.controlServiceServer = controlServiceServer;
+    this.loggingServiceServer = loggingServiceServer;
+    this.retrievalServiceServer = retrievalServiceServer;
+    this.provisioningServiceServer = provisioningServiceServer;
+  }
+
+  @Override
+  public RemoteEnvironment getEnvironment(Environment environment) throws Exception {
+    if (process.isDone()) {
+      DockerContainerEnvironment dockerContainerEnvironment = process.get();
+      checkArgument(
+          environment.getUrl().equals(dockerContainerEnvironment.getEnvironment().getUrl()),
+          "A %s must only be queried for a single %s. Existing %s, Argument %s",
+          SingletonDockerContainerManager.class.getSimpleName(),
+          Environment.class.getSimpleName(),
+          dockerContainerEnvironment.getEnvironment().getUrl(),
+          environment.getUrl());
+    }
+    // TODO: Generate this with an IDGenerator, or any other appropriate source.
+    String environmentId = Long.toString(-123);
+    Path workerPersistentDirectory = Files.createTempDirectory("worker_persistent_directory");
+    Path semiPersistentDirectory = Files.createTempDirectory("semi_persistent_dir");
+    String containerImage = environment.getUrl();
+    String loggingEndpoint = loggingServiceServer.getApiServiceDescriptor().getUrl();
+    String artifactEndpoint = retrievalServiceServer.getApiServiceDescriptor().getUrl();
+    String provisionEndpoint = provisioningServiceServer.getApiServiceDescriptor().getUrl();
+    String controlEndpoint = controlServiceServer.getApiServiceDescriptor().getUrl();
+    ProcessBuilder builder =
+        new ProcessBuilder(
+            "docker",
+            "run",
+            "-v",
+            String.format("%s:%S", workerPersistentDirectory, semiPersistentDirectory),
+            containerImage,
+            String.format("--id=%s", environmentId),
+            String.format("--logging_endpoint=%s", loggingEndpoint),
+            String.format("--artifact_endpoint=%s", artifactEndpoint),
+            String.format("--provision_endpoint=%s", provisionEndpoint),
+            String.format("--control_endpoint=%s", controlEndpoint),
+            String.format("--semi_persist_dir=%s", semiPersistentDirectory));
+
+    Process subprocess = builder.inheritIO().start();
```

Comment:
I don't think we want the docker process to inherit IO streams. At the most, we can maybe share stdout/stderr (if those aren't being captured in any other way).